nibabel/streamlines/compact_list.py

```diff
@@ -0,0 +1,232 @@
+import numpy as np
+
+
+class CompactList(object):
+    """ Class for compacting list of ndarrays with matching shape except for
+    the first dimension.
+    """
+
+    BUFFER_SIZE = 87382*4  # About 4 Mb if item shape is 3 (e.g. 3D points).
+
+    def __init__(self, iterable=None):
+        """
+        Parameters
+        ----------
+        iterable : iterable (optional)
+            If specified, create a ``CompactList`` object initialized from
+            iterable's items. Otherwise, create an empty ``CompactList``.
+
+        Notes
+        -----
+        If `iterable` is a ``CompactList`` object, a view is returned and no
+        memory is allocated. For an actual copy use the `.copy()` method.
+        """
+        # Create new empty `CompactList` object.
+        self._data = np.array(0)
+        self._offsets = np.array([], dtype=int)
+        self._lengths = np.array([], dtype=int)
+
+        if isinstance(iterable, CompactList):
+            # Create a view.
+            self._data = iterable._data
+            self._offsets = iterable._offsets
+            self._lengths = iterable._lengths
+
+        elif iterable is not None:
+            offsets = []
+            lengths = []
+            # Initialize the `CompactList` object from iterable's item.
+            offset = 0
+            for i, e in enumerate(iterable):
+                e = np.asarray(e)
+                if i == 0:
+                    new_shape = (CompactList.BUFFER_SIZE,) + e.shape[1:]
+                    self._data = np.empty(new_shape, dtype=e.dtype)
+
+                end = offset + len(e)
+                if end >= len(self._data):
+                    # Resize needed, adding `len(e)` new items plus some buffer.
+                    nb_points = len(self._data)
+                    nb_points += len(e) + CompactList.BUFFER_SIZE
+                    self._data.resize((nb_points,) + self.shape)
```

Comment:
I'm guessing this is expensive.   If all the elements in the list are arrays, then you can calculate the needed shape in a first pass, and then construct the array.  But I guess you often have something like a generator producing objects where it's expensive to get the shape?